### PR TITLE
Features standard output

### PR DESCRIPTION
**Pull request by sofiia-chorna**
Added a new standard output "features" to return the internal or learned features from a model (model's intermediate values before final predictions). The output format checks were taken from the Plumed [doc](https://www.plumed.org/doc-master/user-doc/html/_m_e_t_a_t_e_n_s_o_r.html).

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [ ] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
 📚 [Download documentation preview for this pull-request](https://nightly.link/lab-cosmo/metatensor/actions/artifacts/1874675914.zip)

### CODE REVIEW

**Comment by frostedoyster** [comment]
It looks good to me! Is this always meant to be used without units? Perhaps it would be worth to add this to the docs

**Comment by sofiia-chorna** [comment]
> It looks good to me! Is this always meant to be used without units? If so, perhaps it would be worth adding it to the docs

Thanks for checking! I added a note regarding the units.

File: python/metatensor-torch/metatensor/torch/atomistic/outputs.py [hunk 1]
```diff
@@ -202,3 +142,107 @@ def _check_energy_like(
...
+    """
+    Ensure the TensorMap has a single block with the expected key
+    """
+    if value.keys != Labels("_", torch.tensor([[0]])):
+        raise ValueError(
+            f"invalid keys for '{name}' output: expected `Labels('_', [[0]])`"
+        )
+
+
+def _validate_components(name: str, block: TensorBlock):
```
Comment (Luthaf) [inline python/metatensor-torch/metatensor/torch/atomistic/outputs.py:241]:
```suggestion
def _validate_no_components(name: str, block: TensorBlock):
```

File: python/metatensor-torch/metatensor/torch/atomistic/outputs.py [hunk 2]
```diff
@@ -202,3 +142,107 @@ def _check_energy_like(
...
+    # Should not have any explicit gradients
+    # all gradient calculations are done using autograd
+    if len(features_block.gradients_list()) > 0:
+        raise ValueError(
+            "invalid gradients for 'features' output: it should not have any explicit",
+            "gradients. all gradient calculations should be done using autograd",
+        )
+
+
+def _validate_samples(
```
Comment (Luthaf) [inline python/metatensor-torch/metatensor/torch/atomistic/outputs.py:177]:
```suggestion
def _validate_atomic_samples(
```

File: docs/src/atomistic/outputs.rst [hunk 3]
```diff
@@ -167,3 +167,43 @@ Energy ensemble gradients
...
+    - same as `Energy`_
+    - same as `Energy`_
+
+  * - samples
+    - same as `Energy`_
+    - same as `Energy`_
+
+  * - components
+    - same as `Energy`_
+    - same as `Energy`_
```
Comment (Luthaf) [inline docs/src/atomistic/outputs.rst:198]: I think it is worth repeating the information here, we used same as `Energy` for the energy ensemble which is related, but this output is not related to the energy and might end up in a different page.

File: docs/src/atomistic/outputs.rst [hunk 4]
```diff
@@ -167,3 +167,43 @@ Energy ensemble gradients
...
 output (see `Energy gradients`_).
+
+.. _features:
+
+Features
+^^^^^^^^
+
+Last layer features represent the outputs generated by the model before the final predictions,
+such as energy or forces, are made. They are associated with the ``"features"`` key in the model
+outputs, and must adhere to the following metadata:
```
Comment (Luthaf) [inline docs/src/atomistic/outputs.rst:178]: These are not necessarily "last layer" features, since there might not even be multiple layer, or another prediction made by the model.

File: docs/src/atomistic/outputs.rst [hunk 5]
```diff
@@ -167,3 +167,43 @@ Energy ensemble gradients
...
+    - same as `Energy`_
+    - same as `Energy`_
+
+  * - properties
+    -
+    - the features can have arbitrary properties.
+
+  * - gradients
+    -
+    - the features should not have any explicit gradients, all gradient calculations are done using autograd.
```
Comment (Luthaf) [inline docs/src/atomistic/outputs.rst:206]: I think we'll want to allow explicit gradients as well, just in case. I'll write this part.

**Review by Luthaf — COMMENTED** [review]
This looks good overall. The only point is about the general name: I'm not sure we want to use "features" for this. A couple of potential alternatives:

- "fingerprint"
- "representation"
- "descriptor"
- "ml-<any of the above>"
- ...

File: docs/src/atomistic/outputs.rst (hunk 3 above)
Comment (sofiia-chorna) [reply docs/src/atomistic/outputs.rst:198]: So, you think we need to split the pages with the standard outputs related and not related to the energy?

**Comment by sofiia-chorna** [comment]
> This looks good overall. The only point is about the general name: I'm not sure we want to use "features" for this. A couple of potential alternatives:
> 
> * "fingerprint"
> * "representation"
> * "descriptor"
> * "ml-"
> * ...

@Luthaf @frostedoyster 
Thanks for reviewing!
I’m not sure which name among proposed is better... Unless you have some preferences, we could ask in the metatensor chat, if you wish.

File: docs/src/atomistic/outputs.rst (hunk 4 above)
Comment (ceriottm) [reply docs/src/atomistic/outputs.rst]: technically these are internal/hidden/learned features. all in all, `features` might be the least bad possible name if we want to have a generic way to refer to the intermediate values in the model. maybe 'learned_features' or 'model_features' but I'm not sure there's much gain from all those extra letters.

Comment (Luthaf) [reply docs/src/atomistic/outputs.rst]: this output is not only intended for internal/learned features, but any kind of representation (so PLUMED would move to use this output instead of the current custom output used over there for CV)

Comment (ceriottm) [reply docs/src/atomistic/outputs.rst]: IMO then `features` is the most generic catch-all term - but I support polling MLDEV.

File: docs/src/atomistic/outputs.rst (hunk 3 above)
Comment (Luthaf) [reply docs/src/atomistic/outputs.rst:198]: We don't need to do it now, but we will likely do something like this at some point. For now just having the documentation for this output be self contained should be enough!

Comment (sofiia-chorna) [reply docs/src/atomistic/outputs.rst]: Ok ok, I added the doc

**Review by frostedoyster — APPROVED** [review]

**Comment by Luthaf** [comment]
I'll update the main description quickly and then this should be good to merge!
